File: 程序员面试金典第6版/面试题 16.07. Maximum LCCI.cpp
class Solution {
public:
    int maximum(int a, int b) {
        uint64_t diff = (uint64_t) a - (uint64_t) b;
        int k = diff >> 63; // c++ 没有循环右移 需要转换成无符号数进行操作
        return a*(k^1) + b*k;
    }
};

File: 程序员面试金典第6版/面试题 16.07. Maximum LCCI.java
// class Solution {
//     public int maximum(int a, int b) {
//         long diff = (long) a - (long) b;
//         int k = (int) (diff >> 63); // long 右移63位得到符号位 正数为0 负数为-1
//         return (k+1)*a - k*b;
//     }
// }

class Solution {
    public int maximum(int a, int b) {
        long diff = (long) a - (long) b;
        int k = (int) (diff >>> 63); // long 循环右移得到符号位 正数为0 负数为1
        return (k^1)*a + k*b;
    }
}
